backoffcache.go

```diff
@@ -0,0 +1,259 @@
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-peerstore/addr"
+)
+
+// BackoffDiscovery is an implementation of discovery that caches peer data and attenuates repeated queries
+type BackoffDiscovery struct {
+	disc         discovery.Discovery
+	stratFactory BackoffFactory
+	peerCache    map[string]*backoffCache
+	peerCacheMux sync.RWMutex
+}
+
+func NewBackoffDiscovery(disc discovery.Discovery, stratFactory BackoffFactory) (discovery.Discovery, error) {
+	return &BackoffDiscovery{
+		disc:         disc,
+		stratFactory: stratFactory,
+		peerCache:    make(map[string]*backoffCache),
+	}, nil
+}
+
+type backoffCache struct {
+	nextDiscover time.Time
+	prevPeers    map[peer.ID]peer.AddrInfo
+
+	peers      map[peer.ID]peer.AddrInfo
+	sendingChs map[chan peer.AddrInfo]int
+
+	ongoing bool
+	strat   BackoffStrategy
+	mux     sync.Mutex
+}
+
+func (d *BackoffDiscovery) Advertise(ctx context.Context, ns string, opts ...discovery.Option) (time.Duration, error) {
+	return d.disc.Advertise(ctx, ns, opts...)
+}
+
+func (d *BackoffDiscovery) FindPeers(ctx context.Context, ns string, opts ...discovery.Option) (<-chan peer.AddrInfo, error) {
+	// Get options
+	var options discovery.Options
+	err := options.Apply(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get cached peers
+	d.peerCacheMux.RLock()
+	c, ok := d.peerCache[ns]
+	d.peerCacheMux.RUnlock()
+
+	/*
+		Overall plan:
+		If it's time to look for peers, look for peers, then return them
+		If it's not time then return cache
+		If it's time to look for peers, but we have already started looking. Get up to speed with ongoing request
+	*/
+
+	// Setup cache if we don't have one yet
+	if !ok {
+		pc := &backoffCache{
+			nextDiscover: time.Time{},
+			prevPeers:    make(map[peer.ID]peer.AddrInfo),
+			peers:        make(map[peer.ID]peer.AddrInfo),
+			sendingChs:   make(map[chan peer.AddrInfo]int),
+			strat:        d.stratFactory(),
+		}
+		d.peerCacheMux.Lock()
+		c, ok = d.peerCache[ns]
+
+		if !ok {
+			d.peerCache[ns] = pc
+			c = pc
+		}
+
+		d.peerCacheMux.Unlock()
+	}
+
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	findPeers := !ok
+	timeExpired := false
+	if !findPeers {
+		timeExpired = time.Now().After(c.nextDiscover)
+		findPeers = timeExpired && !c.ongoing
+	}
+
+	// If we should find peers then setup a dispatcher channel for dispatching incoming peers
+	if findPeers {
+		pch, err := d.disc.FindPeers(ctx, ns, opts...)
+		if err != nil {
+			return nil, err
+		}
+
+		c.ongoing = true
+		go findPeerDispatcher(ctx, c, pch)
+		// If it's not yet time to search again then return cached peers
+	} else if !timeExpired {
+		chLen := options.Limit
+
+		if chLen == 0 {
+			chLen = len(c.prevPeers)
+		} else if chLen > len(c.prevPeers) {
+			chLen = len(c.prevPeers)
+		}
+		pch := make(chan peer.AddrInfo, chLen)
+		for _, ai := range c.prevPeers {
+			pch <- ai
+		}
+		close(pch)
+		return pch, nil
```

Comment:
If we're still in the backoff time we return a closed channel with all the peers we know about. Certainly there are cases where we'd rather hear "there's nothing new" then get the old list again.

Any thoughts on how to differentiate these two functionalities? For example, we could use an implementation specific discovery option (e.g. a string-bool pair in `options.Other`), an instance variable, or just assume we should return "there's nothing new" and add the returning the cache functionality in upon future request.

Reply:
Plan is to just return all of them since there may be multiple consumers of this object. Those consumers can deal with caching and calculating if "there's nothing new"